src/charm.py

```diff
@@ -3,4 +3,82 @@
 # Copyright 2023 Canonical Ltd.
 # See LICENSE file for licensing details.
 
-"""Charm for Synapse."""  # pragma: no cover
+"""Charm for Synapse on kubernetes."""
+
+import logging
+from typing import Any, Dict
+
+import ops
+from ops.main import main
+
+from charm_state import CharmState
+from constants import CHECK_READY_NAME, COMMAND_PATH, SYNAPSE_CONTAINER_NAME, SYNAPSE_SERVICE_NAME
+from exceptions import CharmConfigInvalidError, CommandMigrateConfigError
+from synapse import Synapse
+
+logger = logging.getLogger(__name__)
+
+
+class SynapseCharm(ops.CharmBase):
+    """Charm the service."""
+
+    def __init__(self, *args: Any) -> None:
+        """Construct.
+
+        Args:
+            args: class arguments.
+        """
+        super().__init__(*args)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        try:
+            self._charm_state = CharmState.from_charm(charm=self)
+        except CharmConfigInvalidError as exc:
+            self.model.unit.status = ops.BlockedStatus(exc.msg)
+            return
+        self._synapse = Synapse(charm_state=self._charm_state)
+
+    def _on_config_changed(self, event: ops.HookEvent) -> None:
+        """Handle changed configuration.
+
+        Args:
+            event: Event triggering after config is changed.
+        """
+        container = self.unit.get_container(SYNAPSE_CONTAINER_NAME)
+        if not container.can_connect():
+            event.defer()
+            self.unit.status = ops.WaitingStatus("Waiting for pebble")
+            return
+        self.model.unit.status = ops.MaintenanceStatus("Configuring Synapse")
+        try:
+            self._synapse.execute_migrate_config(container)
+        except CommandMigrateConfigError as exc:
+            self.model.unit.status = ops.BlockedStatus(exc.msg)
+            event.defer()
```

Comment:
I don't think the defer here is necessary since the action of user correcting the configuration will trigger another config-changed event.

Reply:
Changed. :)